src/gba/vfame.c

```diff
@@ -0,0 +1,296 @@
+		break;
+	case 0x1C0000:
+	case 0x1D0000:
+		value = _patternRightShift1(addr) ^ 0xFF00;
+		break;
+	case 0x1E0000:
+	case 0x1F0000:
+		value = _patternRightShift1(addr) ^ 0x00FF;
+		break;
+	}
+
+	return value & 0xFFFF;
+}
+
+static uint32_t _patternRightShift1(uint32_t addr) {
+	return (addr >> 1) & 0xFFFF;
+}
+static uint32_t _patternRightShift2(uint32_t addr) {
+	uint32_t value = addr & 0xFFFF;
+	value = value >> 2;
+	value += (addr & 3) == 2 ? 0x8000 : 0;
+	value += (addr & 0x10000) ? 0x4000 : 0;
+	return value;
+}
+
+void GBAVFameSramWrite(struct GBAVFameCart* cart,uint32_t address, int8_t value, uint8_t* sramData) {
+	// A certain sequence of writes to SRAM FFF8->FFFC can enable or disable "mode change" mode
+	// Currently unknown if these writes have to be sequential, or what happens if you write different values, if anything
+	if (address >= 0x0E00FFF8 && address <= 0xE00FFFC) {
+		cart->writeSequence[address - 0x0E00FFF8] = value;
+		if (address == 0xE00FFFC) {
+			if (memcmp(MODE_CHANGE_START_SEQUENCE, cart->writeSequence, sizeof(MODE_CHANGE_START_SEQUENCE))==0){
+				cart->acceptingModeChange = true;
+			}
+			if (memcmp(MODE_CHANGE_END_SEQUENCE, cart->writeSequence, sizeof(MODE_CHANGE_END_SEQUENCE))==0){
+				cart->acceptingModeChange = false;
+			}
+		}
+	}
+
+	// If we are in "mode change mode" we can change either SRAM or ROM modes
+	// Currently unknown if other SRAM writes in this mode should have any effect
+	if (cart->acceptingModeChange) {
+		if (address == 0x0E00FFFE) {
+			cart->sramMode = value;
+		} else if (address == 0xE00FFFD) {
+			cart->romMode = value;
+		}
+	}
+
+	if (cart->sramMode == -1) {
+		// when SRAM mode is uninitialised you can't write to it
+		return;
+	}
+
+	// if mode has been set - the address and value of the SRAM write will be modified
+	address = _modifySramAddress(cart->cartType, address, cart->sramMode);
+	value = _modifySramValue(cart->cartType, value, cart->sramMode);
+	// these writes are mirrored
+	if (address >= 0xE008000) {
+		address -= 0x8000;
+	}
+	sramData[address & (SIZE_CART_SRAM - 1)] = value;
+	sramData[(address + 0x8000) & (SIZE_CART_SRAM - 1)] = value;
+}
+
+static uint32_t _modifySramAddress(enum GBAVFameCartType type, uint32_t address, int mode) {
+	mode = (mode & 0x3);
+	if (mode == 0) {
+		return address;
+	} else if (type == VFAME_GEORGE) {
+		return _reorderBits(address, ADDRESS_REORDERING_GEORGE[mode-1], 16);
+	} else {
+		return _reorderBits(address, ADDRESS_REORDERING[mode-1], 16);
+	}
+}
+
+static int8_t _modifySramValue(enum GBAVFameCartType type, int8_t value, int mode) {
+	mode = (mode & 0xF) >> 2;
+	if (mode == 0) {
+		return value;
+	} else if (type == VFAME_GEORGE) {
+		return _reorderBits(value, VALUE_REORDERING_GEORGE[mode-1], 8);
+	} else {
+		return _reorderBits(value, VALUE_REORDERING[mode-1], 8);
+	}
+}
+
+// Reorder bits in a byte according to the reordering given
+static int _reorderBits(uint32_t value, const char* reordering, int reorderLength) {
```

Comment:
Although I know my wait state code does this, it's kinda old and should be fixed: this should be const uint8_t*, since you're shifting on it. The possibility of negative shifts really bothers some compilers.